File: Algorithms/0926_Flip_String_to_Monotone_Increasing/Java/FlipStringtoMonotoneIncreasingSolution1.java

// Space: O(n)
// Time: O(n)

class Solution {
    public int minFlipsMonoIncr(String s) {
        if (s.length() == 1) return 0;
        int[] lCache = new int[s.length()];
        int[] rCache = new int[s.length()];
        int res;
        lCache[0] = s.charAt(0) == '0' ? 0 : 1;
        rCache[s.length() - 1] = s.charAt(s.length() - 1) == '1' ? 0 : 1;

        for (int i = 1; i < s.length(); i++) {
            lCache[i] = lCache[i - 1] + s.charAt(i) - '0';
        }

        for (int i = s.length() - 2; i >= 0; i--) {
            rCache[i] = rCache[i + 1] + '1' - s.charAt(i);
        }

        res = Math.min(lCache[s.length() - 1], rCache[0]);

        for (int i = 0; i < s.length() - 1; i++) {
            res = Math.min(res, lCache[i] + rCache[i + 1]);
        }
        // use following loop is also correct
        // for (int i = 1; i < s.length(); i++) {
        //     res = Math.min(res, lCache[i - 1] + rCache[i]);
        // }

        return res;
    }
}

File: Algorithms/0926_Flip_String_to_Monotone_Increasing/Python/Flip_String_to_Monotone_Increasing_Solution_1.py

# Space: O(n)
# Time: O(n)

class Solution:
    def minFlipsMonoIncr(self, s):
        length = len(s)
        if length <= 1: return 0
        res = 0
        cache = list(map(lambda x: int(x), list(s)))
        l_cache = [0 for _ in range(length)]
        r_cache = [0 for _ in range(length)]

        l_cache[0] = cache[0]
        r_cache[-1] = 1 if cache[-1] == 0 else 0

        for i in range(1, length):
            l_cache[i] = cache[i] + l_cache[i - 1]

        for i in range(length - 2, -1, -1):
            if cache[i] == 1:
                r_cache[i] = r_cache[i + 1]
            else:
                r_cache[i] = r_cache[i + 1] + 1

        res = min(l_cache[-1], r_cache[0])

        for i in range(0, length - 1):
            res = min(res, l_cache[i] + r_cache[i + 1])

        # use following loop is also correct
        # for i in range(1, length):
        #     res = min(res, l_cache[i - 1] + r_cache[i])

        return res




